### PR TITLE
[Core] Refactor Worker and ModelRunner to consolidate control plane communication

### PR DESCRIPTION
**Pull request by stephanie-wang**
Currently the ModelRunner class contains both model execution code and multi-GPU control plane communication code, i.e.`broadcast_tensor_dict` calls. This makes it difficult to swap out the control plane mechanism, e.g., using NCCL vs CPU-based serialization to move the inputs from the LLMEngine to the Workers. It also makes it difficult to improve performance for multi-GPU settings, because we need to reason about all of the possible broadcast calls that may happen during one execution request.

Previously, the Worker and ModelRunner class looked something like this:
```python
class Worker:
  def execute_model(self, execute_model_req=None):
    # Do some broadcast here.
    ...
    return self.model_runner.execute_model(execute_model_req)

class ModelRunner:
  def execute_model(self, execute_model_req=None):
    # Do some more broadcast here.
    ...
    return model_executable(...)
```
This PR refactors the Worker and ModelRunner classes to consolidate all of the control plane communication into a central location. This adds a new class `ModelInput`, which acts as the interface between `Worker` and `ModelRunner`. The contract is that ExecuteModelRequest contains CPU-only metadata, while `ModelInput`'s tensors should already be on the correct device. Now, the Worker class looks like this:
```python
class Worker:
  def execute_model(self, execute_model_req=None):
    # *All* broadcast / other communication happens in this function.
    model_input: ModelInput = self.prepare_model_input(execute_model_req)
    # Worker.execute_model_local and ModelRunner code contain single-device code only.
    return self.execute_model_local(model_input)
```
This refactor has a few advantages:
- Reduces the number of broadcast_tensor_dict calls, which can save some latency
- Cleaner separation between control plane communication vs. single-GPU logic. Each ModelRunner needs to explicitly state what inputs it requires by defining a ModelInput (subclass). This requires a bit more developer effort but should make it easier to try out other optimizations in the future:
  - Other control plane implementations (e.g., CPU-based serialization instead of NCCL)
  - Moving the rank 0 worker to a different process from the LLMEngine.
  - Pipelining model execution with metadata movement between CPU and GPU

One complication is that the speculative decoding code goes back and forth between ExecuteModelRequest and ModelInput, whereas other workers only convert from ExecuteModelRequest to ModelInput. Thus, for the speculative decoding path, it's easier for now to keep the per-step broadcast. These extra k broadcasts could also be consolidated in the future, by either supporting ModelInput -> ExecuteModelRequest, or by making it possible to modify a ModelInput.

TODO:
- [x] Update neuron model runner
- [x] Disallow control plane communication ops inside ModelRunner code

### CODE REVIEW

**Comment by youkaichao** [comment]
Thanks, Stephanie! FYI I have a PR with somewhat similar functionality https://github.com/vllm-project/vllm/pull/5354 .

**Comment by stephanie-wang** [comment]
> Thanks, Stephanie! FYI I have a PR with somewhat similar functionality #5354 .

Ah I see, thanks! Do you want to take a quick look at this PR and see what you think? I think this is a superset of #5354.

**Comment by youkaichao** [comment]
Will take a look! I like this idea:

> The contract is that ExecuteModelRequest contains CPU-only metadata, while ModelInput's tensors should already be on the correct device.

**Review by rkooo567 — COMMENTED** [review]
Looks pretty clean! A little worried about spec decoding case, but I guess there must be a solution. 

IIUC, we cannot run tests with execute_model_local with this PR right? Are we planning to add test in the future PRs?

File: vllm/sequence.py [hunk 1]
```diff
@@ -867,3 +877,138 @@ def clone(
             num_lookahead_slots=self.num_lookahead_slots,
             running_queue_size=self.running_queue_size,
         )
+
+
+@dataclasses.dataclass(frozen=True)
+class ModelInput:
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply vllm/sequence.py:883]: Sounds good! I can create a base empty `ModelInput` class and then subclass with GPUModelInput, since that one has a bunch of shared fields with others.

**Comment by rkooo567** [comment]
added action-required flag (lmk when it is ready again), and cc @cadedaniel

**Comment by stephanie-wang** [comment]
> added action-required flag (lmk when it is ready again), and cc @cadedaniel. I think after my comments are addressed, tests pass, I am pretty happy with a PR now

Thanks for the review! I need to rerun TP benchmarks but the code is ready for review again.

**Review by rkooo567 — COMMENTED** [review]
Code looks good! I guess we need a bit more work to make tests pass?

**Comment by stephanie-wang** [comment]
Ran a couple benchmarks on this PR. On A100s, TP=4 2.8% and TP=8 2% per-token latency improvement (about 200-300us) respectively. This looks consistent with results from #5354.

**Comment by rkooo567** [comment]
spec decoding test failure is unrelated right? Since we don't switch to non-nccl backend yet.

**Comment by stephanie-wang** [comment]
> spec decoding test failure is unexpected right? Since we don't switch to non-nccl backend yet.

Yeah just a bug, should be fixed now.

**Review by rkooo567 — APPROVED** [review]

**Review by cadedaniel — APPROVED** [review]
did a light pass. thanks for the consolidation. glad it already brings a small latency win.

File: vllm/sequence.py [hunk 2]
```diff
@@ -163,6 +167,8 @@ def get_num_computed_tokens(self) -> int:
 
     def update_num_computed_tokens(self, num_new_computed_tokens: int):
         """Update number of tokens computed so far."""
+        # TODO: Check who calls this and make sure it's synchronized across
+        # driver and workers.
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply vllm/sequence.py:171]: Oops, thanks, will remove.

File: vllm/model_input.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
+"""Worker-local model inputs. These define the inputs to different model
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply vllm/model_input.py:1]: Good idea, will move.

**Comment by youkaichao** [comment]
Will take a detailed look in the next week! The overall idea of this PR looks good to me. Thanks for the efforts!

File: vllm/sequence.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
     from vllm.spec_decode.metrics import SpecDecodeWorkerMetrics
 
 
-@dataclass
+@dataclasses.dataclass
```
Comment (youkaichao) [inline vllm/sequence.py:21]: why is this change needed? can we reduce the number of lines changed to make the review easier?

Comment (youkaichao) [inline vllm/sequence.py:21]: why is this change needed? can we reduce the number of lines changed to make the review easier?

**Review by youkaichao — CHANGES_REQUESTED** [review]

File: vllm/worker/worker.py [hunk 5]
```diff
@@ -263,37 +252,63 @@ def execute_model(
...
+                        model_input.blocks_to_copy)
 
         # If there is no input, we don't need to execute the model.
-        if num_seq_groups == 0:
+        if model_input.num_seq_groups == 0:
             return []
 
-        output = self.model_runner.execute_model(seq_group_metadata_list,
-                                                 self.gpu_cache)
+        output = self.model_runner.execute_model(model_input, self.gpu_cache)
```
Comment (youkaichao) [inline vllm/worker/worker.py:306]: this function is decorated with `@disable_communication`, then what will happen when we have TP enabled?

File: vllm/sequence.py (hunk 4 above)
Comment (stephanie-wang) [reply vllm/sequence.py:21]: Ah my bad, I think this was because `dataclasses.field` had a conflict in an earlier version with a local variable called `field`. Will revert the changes to this file.

File: vllm/worker/worker.py (hunk 5 above)
Comment (stephanie-wang) [reply vllm/worker/worker.py:306]: This is only referring to control plane communication, so it should not affect TP execution (unless that is somehow calling broadcast_tensor_dict).

Comment (youkaichao) [reply vllm/worker/worker.py:306]: This is kind of misleading. I don't think we need to enforce it with such a heavy mechanism (and also slight runtime overhead). Just keep it in docstring is okay. Very few people will touch these functions or call `broadcast_tensor_dict` themselves.

File: vllm/worker/model_input.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    This base class contains metadata needed for the base model forward pass
+    but not metadata for possible additional steps, e.g., sampling. Model
+    runners that run additional steps should subclass this method to add
+    additional fields.
+    """
+    num_seq_groups: Optional[int] = None
+    blocks_to_swap_in: Optional[torch.Tensor] = None
+    blocks_to_swap_out: Optional[torch.Tensor] = None
+    blocks_to_copy: Optional[torch.Tensor] = None
+
```
Comment (youkaichao) [inline vllm/worker/model_input.py:207]: I feel this is abstraction leakage. Ideally we should have `WorkerInput` and `ModelInput`, and `Worker` should not peak into `ModelInput`.

File: vllm/worker/worker.py (hunk 5 above)
Comment (stephanie-wang) [reply vllm/worker/worker.py:306]: How about renaming it to `forbid_control_plane_communication` or `worker_local`? And can you say more what your concerns are? The run-time overhead for the check is negligible.

Comment (youkaichao) [reply vllm/worker/worker.py:306]: I think it is fine to just document that we don't call `broadcast_tensor_dict` in this function. We don't need to enforce it with runtime check.

Actually, after the refactor, if we don't call `broadcast_tensor_dict` in this function, and this function only deals with input that is already broadcast, who will bother adding a `broadcast_tensor_dict` here?

That's why I think this `@disable_communication` is unnecessary.

**Comment by stephanie-wang** [comment]
@youkaichao this is ready for another review. Thanks!

**Comment by youkaichao** [comment]
will take a look tmr.

**Comment by stephanie-wang** [comment]
> @stephanie-wang can we use the class-in-class style in [#5552 (comment)](https://github.com/vllm-project/vllm/issues/5552#issuecomment-2168825615) ?
> 
> The reason why I want this, is because hardware specific code can be isolated. Currently you have different ModelInput code for different hardware, but **in the same file**, which can cause potential trouble and their conflict. It is possible that some hardware backend needs to import something specific to that platform. By using the class-in-class style, we will never import anything from a hardware backend that is not used.

Sure, will do!

**Comment by youkaichao** [comment]
Looks better to me now. can we rename `WorkerInput` as `WorkerInputBase`, and put it in `worker_base.py`?

Likewise, can we rename `ModelInput` as `ModelRunnerInputBase`, and put it in `model_runner_base.py`?

File: vllm/worker/model_runner_base.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+if TYPE_CHECKING:
+    from vllm.attention import AttentionMetadata
+    from vllm.attention.backends.abstract import AttentionBackend
+    from vllm.model_executor import SamplingMetadata
+
+T = TypeVar('T', bound="ModelInput")
+
+
+@dataclasses.dataclass(frozen=True)
+class ModelInput:
```
Comment (youkaichao) [inline vllm/worker/model_runner_base.py:19]: because we previously use `ModelInput` for GPU input, let's rename this as `ModelInputBase`, so that people will not get confused in the future.

File: vllm/worker/model_runner_base.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+            attn_metadata = attn_backend.make_metadata(**valid_attn_kwargs)
+        if attn_metadata is not None:
+            kwargs["attn_metadata"] = attn_metadata
+        return kwargs
+
+    @staticmethod
+    def _init_sampling_metadata_from_kwargs(  # type: ignore
+            selected_token_indices: Optional[torch.Tensor] = None,
+            sampling_metadata: Optional["SamplingMetadata"] = None,
```
Comment (youkaichao) [inline vllm/worker/model_runner_base.py:129]: why sampling metadata and attention metadata leak here?

**Comment by stephanie-wang** [comment]
> Looks better to me now. can we rename `WorkerInput` as `WorkerInputBase`, and put it in `worker_base.py`?
> 
> Likewise, can we rename `ModelInput` as `ModelRunnerInputBase`, and put it in `model_runner_base.py`?

Renamed ModelInput to ModelRunnerInputBase, but I left WorkerInput as WorkerInput because it is shared by all the Worker subclasses (they all basically use the same few fields).

File: vllm/worker/model_runner_base.py (hunk 8 above)
Comment (stephanie-wang) [reply vllm/worker/model_runner_base.py]: Moved these to helper methods.

File: vllm/worker/model_runner_base.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    """
+
+    @classmethod
+    @abstractmethod
+    def new(cls: Type[T], **kwargs) -> T:
+        """
+        Create a new instance of this class. Populate the new instance with
+        the given kwargs.
+        """
+        raise NotImplementedError
```
Comment (youkaichao) [inline vllm/worker/model_runner_base.py:112]: why do we create a `new` method, rather than directly using `__init__`?

File: vllm/worker/model_runner_base.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        Attributes whose values are None will not be added to the returned
+        dictionary.
+        """
+        tensor_dict: Dict[str, Union[int, torch.Tensor]] = {}
+        for attr in attrs:
+            val = getattr(self, attr, None)
+            if val is not None:
+                tensor_dict[attr] = val
+
+        return tensor_dict
```
Comment (youkaichao) [inline vllm/worker/model_runner_base.py:140]: IMHO, if an attribute is `None`, we can also put it in the returning dict. We don't need to have a special logic for this.

File: vllm/worker/model_runner.py [hunk 11]
```diff
@@ -39,40 +48,80 @@
...
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        tensor_dict = self._get_attrs([
+            "input_tokens",
+            "input_positions",
+            "lora_requests",
+            "lora_mapping",
+            "multi_modal_kwargs",
+        ])
+        _add_attn_metadata_broadcastable_dict(tensor_dict, self.attn_metadata)
+        return tensor_dict
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:81]: I would say, this is less intuitive than:

```python
tensor_dict = {
    "input_tokens": self.input_tokens,
    ...
}
tensor_dict.update(self.attn_metadata.asdict_zerocopy())
```

File: vllm/worker/model_runner.py [hunk 12]
```diff
@@ -39,40 +48,80 @@
...
+    sampling_metadata: Optional["SamplingMetadata"] = None
+    # Used for speculative decoding. We do not broadcast it because it is only
+    # used by the driver worker.
+    is_prompt: Optional[bool] = None
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        tensor_dict = super().as_broadcastable_tensor_dict()
+        _add_sampling_metadata_broadcastable_dict(tensor_dict,
+                                                  self.sampling_metadata)
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:107]: inheritance does not give any benefit here, but rather decreases the readability. same as above comments, we can just expand the code, write out these fields directly.

File: vllm/worker/model_runner_base.py (hunk 9 above)
Comment (stephanie-wang) [reply vllm/worker/model_runner_base.py:112]: The same broadcasted kwargs are deserialized to both WorkerInput and ModelInput dataclasses, so we need to customize the `__init__` to only take the kwargs that are valid for the current dataclass.

File: vllm/worker/model_runner.py (hunk 11 above)
Comment (stephanie-wang) [reply vllm/worker/model_runner.py:81]: `_get_attrs` is used to filter out Nones. I think this is quite subjective; IMO, `self._get_attrs` is clear enough.

File: vllm/worker/model_runner_base.py (hunk 10 above)
Comment (youkaichao) [reply vllm/worker/model_runner_base.py:140]: `broadcast_object ` can accept value of `None`. All non-tensor parts will be squeezed into a python object, without additional `broadcast` call.

File: vllm/worker/model_runner.py (hunk 12 above)
Comment (stephanie-wang) [reply vllm/worker/model_runner.py:107]: The benefit of inheritance here is that if the broadcasted fields change, we only need to modify them in one place. Otherwise, a developer can mistakenly add it in only one place, type checking will not complain, but the broadcast will be buggy for all subclasses.

Do you prefer to duplicate the code? I disagree in this case but I don't feel strongly about it.

File: vllm/worker/model_runner_base.py (hunk 10 above)
Comment (stephanie-wang) [reply vllm/worker/model_runner_base.py:140]: But there is still serialization cost, no?

Comment (youkaichao) [reply vllm/worker/model_runner_base.py:140]: just a few more bytes in the serialized bytes, I don't think it is worth optimizing.

File: vllm/worker/model_runner.py (hunk 12 above)
Comment (youkaichao) [reply vllm/worker/model_runner.py:107]: I think duplicating the code here is fine. When someone modify them in one place, the worker side will have a code to consume that field. Although type checkers cannot ccomplain, test code will catch it.

File: vllm/worker/worker_base.py [hunk 13]
```diff
@@ -98,6 +114,169 @@ def list_loras(self) -> Set[int]:
...
+
+    @property
+    @abstractmethod
+    def model_runner(self) -> ModelRunnerBase:
+        """
+        Get the worker's model runner. Used by the default `execute_model`. If
+        the worker's model runner does not follow the ModelRunnerBase
+        interface, then inherit from WorkerBase instead.
+        """
+        raise NotImplementedError
```
Comment (youkaichao) [inline vllm/worker/worker_base.py:199]: why do we need this? originally we already have `.model_runner` attribute.

**Review by youkaichao — APPROVED** [review]
LGTM, thanks for addressing all my comments!
